README
-----------------------------------------------------------------------
               The Duchamp Source Finder
-----------------------------------------------------------------------
Duchamp 1.6.1 -- an object finder for spectral-line data cubes
Copyright (C) 2006-2014
Matthew Whiting, CSIRO Astronomy & Space Science

Duchamp is free software; you can redistribute it and/or modify it
under the terms of the GNU General Public License as published by the
Free Software Foundation; either version 2 of the License, or (at your
option) any later version.

Duchamp is distributed in the hope that it will be useful, but WITHOUT
ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
for more details.

You should have received a copy of the GNU General Public License
along with Duchamp; if not, write to the Free Software Foundation,
Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307, USA

Correspondence concerning Duchamp may be directed to:
   Internet email: Matthew.Whiting [at] atnf.csiro.au
   Postal address: Dr. Matthew Whiting
                   Australia Telescope National Facility, CSIRO
                   PO Box 76
                   Epping NSW 1710
                   AUSTRALIA
-----------------------------------------------------------------------

Introduction
============

Duchamp is a stand-alone program designed to find objects in
astronomical data cubes, particularly spectral-line observations. Its
features include a wavelet-based reconstruction technique for reducing
the noise in the cube (and thereby enhancing detectability of
sources), easy-to-use text-based interface, flexibility to control all
relevant parameters such as detection thresholds, and a useful range
of text- and graphics-based output.

Duchamp works on any FITS image using the CFITSIO package, and uses
Mark Calabretta's WCSLIB library to provide accurate position and
velocity information for all detected sources.


Obtaining and Building Duchamp
==============================

The Duchamp web page is at 
http://www.atnf.csiro.au/people/Matthew.Whiting/Duchamp
where you can download a gzipped tar archive of the source code.

Duchamp uses three main external libraries: pgplot (although see note
below), cfitsio (version 2.5 and greater, version 3+ preferred) and
wcslib. If you do not have the libraries, they can be downloaded from
the following locations:

PGPLOT -- http://www.astro.caltech.edu/~tjp/pgplot/ 
cfitsio -- http://heasarc.gsfc.nasa.gov/docs/software/fitsio/fitsio.html 
wcslib -- http://www.atnf.csiro.au/people/Mark.Calabretta/WCS/index.html

Note that PGPLOT is optional. Duchamp can be compiled and run without
PGPLOT, but you will miss out on the useful graphical outputs!  The
other two libraries, cftisio and wcslib, are essential.


Basic Compilation and Installation
----------------------------------

Duchamp can be built and installed on Unix systems by typing (note
that the terminal prompt here is represented by > -- don't type this
character!):

 >  ./configure
 >  make
 >  make lib   (optional -- this makes libraries for use in code
    	        development)
 >  make clean (again optional -- to remove the object files from the
    	        src directory)
 >  make install 

This default setup will search in standard locations for the necessary
libraries, and install the executable ("Duchamp") in /usr/local/bin (a
copy will also be in the current directory). It also installs the
libraries libduchamp.1.6.a and libduchamp.1.6.so in /usr/local/lib and
the header (.hh) files in /usr/local/include/duchamp. If you want
these to go somewhere else, eg if you don't have write-access to that
directory, or you need to tweak the libraries, see the next
section. Otherwise, jump to the testing section.


Tweaking the installation setup
-------------------------------

To install Duchamp in a directory other than /usr/local/bin, use the
--prefix option with configure, specifying the directory above the
\texttt{bin/} directory. eg:

 >  ./configure --prefix=/home/mduchamp

which, if you run "make" and "make install", will result in the binary
being put in the directory /home/mduchamp/bin. Similarly the library
will be put in /home/mduchamp/lib and the header files in
/home/mduchamp/include/duchamp and subdirectories thereof.

If the above-mentioned libraries have been installed in non-standard
locations, or you have more than one version installed on your system,
you can specify specific locations by using the --with-cfitsio=<dir>,
--with-wcslib=<dir> or --with-pgplot=<dir> flags. For example:

 >  ./configure --with-wcslib=/home/mduchamp/wcslib-4.2 

And then just run make in the usual fashion:

 >  make

Duchamp can be compiled without PGPLOT if it is not installed on your
system -- the searching and text-based output remains the same, but
you will not have any graphical output. 
To manually specify this option, use the --without-pgplot flag, eg:

 >  ./configure --without-pgplot

(Note that CFITSIO and WCSLIB are essential, however, so
--without-wcslib or --without-cfitsio will not work.).
Even if you do not give the --without-pgplot option, and the PGPLOT
library is not found, Duchamp will still compile (albeit without
graphical capabilities).

Troubleshooting the installation
--------------------------------

There may be issues with the configure/make/install process that arise 
from non-standard locations of key libraries. The User's Guide has a 
section in Appendix A discussing some of the more common problems 
and potential solutions.

Feeback is welcome if you have trouble getting Duchamp working -- see 
the feedback section below.


Testing
=======

There is a script included in the distribution that allows you to make
sure Duchamp is running correctly. It will use a dummy FITS image in
the verification/ directory -- this image has some Gaussian random
noise, with five Gaussian sources present, plus a dummy WCS. The
script runs Duchamp on this image with three different sets of inputs,
and compares to known results, looking for differences and reporting
any. There should be none reported if everything is working
correctly. To run, enter the command

 >  ./VerifyDuchamp.sh

This performs basic checks on the output files, but ignoring most of
the actual values of source parameters (to avoid picking up just
differences due to precision errors). For complete checks of the
files, run 

 > ./VerifyDuchamp.sh -f

Be warned that on some systems this could provide a large number of
apparent errors which may only be due to precision differences.

Once you know it is working, you can install Duchamp on your system
with the command 

 > make install

(this may need to be run as sudo depending on your system setup and
your prefix directory).

You can also use the dummy image for your own testing if you like (for
instance, testing different thresholds to get a feel for how the
program works). 


Using Duchamp
=============

There are two possible ways to run Duchamp. The first is:

 >  Duchamp -f image.fits

where image.fits is the data cube to be searched. This method simply
uses the default values of all parameters.

The second method allows some determination of the parameter values by
the user. Type:

 >  Duchamp -p parameterFile 

where parameterFile is a file with the input parameters, including the
name of the cube you want to search. There are two example input files
included with the distribution. The smaller one, InputExample, shows
the typical parameters one might want to set. The large one,
InputComplete, lists all possible parameters that can be entered, and
a brief description of them. To get going quickly, just replace the
"your-file-here" in InputExample with your image name, and type 

 >  Duchamp -p InputExample

By default, a map of detections is displayed in a PGPLOT
X-window. This can be disabled by using the flagXOutput parameter, or
using the -x command-line option along with the -f or -p options. The
-x option will override the setting in the parameter file.

You can specify a flux threshold from the command line by using the -t
option. This is particularly useful when combined with -f to enable a
quick search to some flux level:

 >  Duchamp -f image.fits -t 0.001

A User's Guide in the docs/ directory provides complete
documentation. It comes in both postscript and portable document
format (pdf -- note that this contains hyperlinks). This guide will
provide full descriptions of all parameters, and of all steps in the
execution of Duchamp.

Feedback
========

Any questions, please contact me. To report problems or bugs, or to
suggest improvements, please go to the Duchamp Trac wiki site:
http://svn.atnf.csiro.au/trac/duchamp/newticket 
and submit a "ticket", or view previously submitted reports.

Acknowledging the use of Duchamp
================================

Duchamp is provided in the hope that it will be useful for your
research. If you find that it is, I would ask that you acknowledge it
in your publication by using the following:
"This research made use of the Duchamp source finder, produced at
the Australia Telescope National Facility, CSIRO, by M. Whiting."

The journal paper describing Duchamp, and providing basic comparative
tests of its different functions, is Whiting 2012, MNRAS 421,
3242, which can be found online at 
http://onlinelibrary.wiley.com/doi/10.1111/j.1365-2966.2012.20548.x/full.
This paper should be cited when describing Duchamp. 

An earlier conference proceedings paper briefly describing it was:
"Astronomers! Do you know where your galaxies are?", M.Whiting, in
"Galaxies in the Local Volume", Eds. B.Koribalski & H.Jerjen,
Astrophysics & Space Science Proceedings, Springer, 2008, p.343-344
http://adsabs.harvard.edu/abs/2008glv..book..343W

Those interested in the development of a source-finder for ASKAP
(known for now as Selavy), which make use of the Duchamp code, should
read Whiting & Humphreys 2012, PASA 29, 371, which can be found online
at http://www.publish.csiro.au/paper/AS12028.htm

-------
Author: 
  Matthew Whiting, Australia Telescope National Facility, April 2014
  Matthew.Whiting [at] csiro.au



CHANGES
Duchamp version 1.6.1 (2014/05/01)
----------------------------------

Principal reason for the release is to fix a memory leak. This was
occuring when making blank pixel masks, particularly in the
parameterisation step. It was apparent with large files and large
numbers of sources. Thanks to Quang Nguyen for raising the bug and
helping with the identification of the problem.

For those interested in the code base itself, all boolean masks have
been moved to vector<bool> types, with the consequent change of
interface for some functions (including the statistics functions,
Gaussian smoothing, and some spectral utility functions).

A few other potential memory leaks were identified and fixed.

To aid the debugging, the progress bar now also reports the percentage
complete as well as the set of hashes.


Duchamp version 1.6 (2014/04/22)
--------------------------------

Improvements and bug fixes:

* When searching for negative features while using baseline
  subtraction, the parameterisation of the resulting detections was
  not being done correctly. Duchamp now correctly identifies the peak
  absorption location and measures the peak as being the peak
  absorption from the fitted continuum level. A future release will
  add more absorption-line-oriented parameters.

* The way the borders of detections are displayed in annotation files
  has been improved. Duchamp now writes them in a polygon or
  join-the-dots fashion. This reduces the size of the annotation
  files, and is now compatible with the CASA viewer.

* The VOTables now refer to version 1.3 of the VOTable schema. Note
  that they still use the old COOSYS method of specifying the
  coordinates. Moving to the preferred STC method will happen in a
  future release.

* There were some problems with the storage and recovery of parameters
  in the binary catalogue - these have been fixed.

* The BITPIX keyword in the mask FITS files that Duchamp writes are
  now SHORT_IMG, although they can be bigger if warranted by the
  number of objects.

* Duchamp can now read .fits.gz files, and have the output FITS files
  appropriately named in the automatic way (the .gz suffix was not
  being recognised & removed properly).

* Some UCDs in the VOTable were missing semicolons between elements.

* When calculating the cube statistics, Duchamp makes a mask showing
  the locations of blank pixels, flagged channels, and the statistics
  subsection are, so that only the required pixels are used in the
  calculations. If none of these apply, however (that is, if there are
  no blank pixels or flagged channels, and the user is not requesting
  a particular stats subsection via the statsec parameter), then the
  mask is not created. This saves additional memory allocation, which
  can be important for large cubes. Thanks to Andrew Walsh for
  identifying this problem.

* Duchamp exits better (ie. without throwing a segmentation fault)
  when the requested FITS file does not exist.

* The reference cube used for the verification script has been changed
  - the sources therein are now smaller spatially (more appropriately
  so for the pixel size), which gives a better idea as to the
  performance.

Documentation improvmements

* The InputComplete file is more up-to-date, and has information on
  the type and default values of each possible input parameter (this
  information is also reflected in the User Guide). This is in
  preparation for a parameter-editing GUI that will be released in a
  forthcoming version.

* The User's Guide now has a section on common installation issues,
  based on feedback and experience with new users.

Build changes:

* Duchamp is now linked with the C++ compiler. Accordingly, the
  configure script searches for libgfortran when the gfortran compiler
  is specified as the F77 compiler (ie. it has been used to build
  pgplot).

* To allow Duchamp to build on Mac OS X 10.9 (Mavericks), configure
  searches for the libc++ library as well as stdc++. Duchamp has been
  successfully built on 10.9, with fresh installs of pgplot (5.2),
  wcslib (4.20) and cfitsio (3.36), but, as always, your mileage may
  vary. Please contact me via the Trac tickets if you encounter
  problems.

* A number of warnings that came to light with recent XCode upgrades
  on the Mac have been fixed, as well as those from an upgrade to
  gcc-4.4.5 on debian.


Code base changes:

* There are new wcs-pix conversion functions in fitsHeader that allow
  the specification of a single set of pixel and world coordinates
  (rather than arrays).

* There is cleaner separation of the writing functions in the
  VOTableCatalogueWriter class.

* The management of the wavelet filters has been improved, with new
  classes to handle the different filter types, and a factory class to
  generate them. The interface via the Param class is largely the
  same, however.

* The annotation file changes noted above involve a new PixelInfo
  class called Line, with associated files. The method for drawing the
  borders has been moved into its own file (drawBorders.cc) to allow
  dependence on the pgplot library.

* The function Cube::reInvert() has been removed - use Cube::invert()
  instead. Note that the interface for the latter now has two optional
  bool parameters.

* There are improved front-end functions for the writing of arrays to
  the various FITS files.



Duchamp version 1.5 (2013/08/29)
--------------------------------

New features & algorithms:

* There is now a more flexible way to specify channels that should be
  ignored by the source-finding. Use flaggedChannels to provide a list
  of single channels or channel ranges that should be ignored. This
  replaces the flagMW/minMW/maxMW parameters, extending their
  functionality and allowing more than one set of channels to be
  specified. 

* There is now an alternative way to find (and remove) spectral
  baselines, using a median baselining algorithm. This calculates the

src/duchamp.cc
// -----------------------------------------------------------------------
// duchamp.cc: Warning and Error messages.
// -----------------------------------------------------------------------
// Copyright (C) 2006, Matthew Whiting, ATNF
//
// This program is free software; you can redistribute it and/or modify it
// under the terms of the GNU General Public License as published by the
// Free Software Foundation; either version 2 of the License, or (at your
// option) any later version.
//
// Duchamp is distributed in the hope that it will be useful, but WITHOUT
// ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
// FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
// for more details.
//
// You should have received a copy of the GNU General Public License
// along with Duchamp; if not, write to the Free Software Foundation,
// Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307, USA
//
// Correspondence concerning Duchamp may be directed to:
//    Internet email: Matthew.Whiting [at] atnf.csiro.au
//    Postal address: Dr. Matthew Whiting
//                    Australia Telescope National Facility, CSIRO
//                    PO Box 76
//                    Epping NSW 1710
//                    AUSTRALIA
// -----------------------------------------------------------------------
#include <iostream>
#include <stdexcept>
#include <iomanip>
#include <string>
#include <duchamp/duchamp.hh>
#include <fitsio.h>

namespace duchamp
{


  DuchampError::DuchampError(const std::string& message)
    : std::runtime_error(message)
  {}

  DuchampError::~DuchampError() throw()
  {}


  void duchampFITSerror(int status, std::string subroutine, std::string error)
  {
    if(status){
      DUCHAMPWARN(subroutine,error);
      fits_report_error(stderr, status);
    }
  }


}
